File: code/packages/typescript/state-machine/tests/minimize.test.ts
/**
 * Tests for DFA minimization (Hopcroft's algorithm).
 */

import { describe, expect, it } from "vitest";
import { DFA } from "../src/dfa.js";
import { minimize } from "../src/minimize.js";
import { NFA } from "../src/nfa.js";
import { transitionKey } from "../src/types.js";

describe("Minimize Basic", () => {
  it("should not reduce an already minimal DFA", () => {
    const dfa = new DFA(
      new Set(["q0", "q1"]),
      new Set(["a", "b"]),
      new Map([
        [transitionKey("q0", "a"), "q1"],
        [transitionKey("q0", "b"), "q0"],
        [transitionKey("q1", "a"), "q0"],
        [transitionKey("q1", "b"), "q1"],
      ]),
      "q0",
      new Set(["q1"]),
    );
    const minimized = minimize(dfa);
    expect(minimized.states.size).toBe(2);
  });

  it("should merge equivalent states", () => {
    // q1 and q2 are both accepting with identical transitions (self-loops)
    const dfa = new DFA(
      new Set(["q0", "q1", "q2"]),
      new Set(["a", "b"]),
      new Map([
        [transitionKey("q0", "a"), "q1"],
        [transitionKey("q0", "b"), "q2"],
        [transitionKey("q1", "a"), "q1"],
        [transitionKey("q1", "b"), "q1"],
        [transitionKey("q2", "a"), "q2"],
        [transitionKey("q2", "b"), "q2"],
      ]),
      "q0",
      new Set(["q1", "q2"]),
    );
    const minimized = minimize(dfa);
    expect(minimized.states.size).toBe(2);
  });

  it("should remove unreachable states", () => {
    const dfa = new DFA(
      new Set(["q0", "q1", "q_dead"]),
      new Set(["a"]),
      new Map([
        [transitionKey("q0", "a"), "q1"],
        [transitionKey("q1", "a"), "q0"],
        [transitionKey("q_dead", "a"), "q_dead"],
      ]),
      "q0",
      new Set(["q1"]),
    );
    const minimized = minimize(dfa);
    expect(minimized.states.size).toBe(2);
  });

  it("should preserve the language", () => {
    const dfa = new DFA(
      new Set(["q0", "q1", "q2", "q3"]),
      new Set(["a", "b"]),
      new Map([
        [transitionKey("q0", "a"), "q1"],
        [transitionKey("q0", "b"), "q2"],
        [transitionKey("q1", "a"), "q3"],
        [transitionKey("q1", "b"), "q3"],
        [transitionKey("q2", "a"), "q3"],
        [transitionKey("q2", "b"), "q3"],
        [transitionKey("q3", "a"), "q3"],
        [transitionKey("q3", "b"), "q3"],
      ]),
      "q0",
      new Set(["q1", "q2"]),
    );
    const minimized = minimize(dfa);

    const testInputs = [["a"], ["b"], ["a", "a"], ["a", "b"], ["b", "a"], []];
    for (const events of testInputs) {
      expect(dfa.accepts(events)).toBe(minimized.accepts(events));
    }
  });

  it("should handle single-state DFA", () => {
    const dfa = new DFA(
      new Set(["q0"]),
      new Set(["a"]),
      new Map([[transitionKey("q0", "a"), "q0"]]),
      "q0",
      new Set(["q0"]),
    );
    const minimized = minimize(dfa);
    expect(minimized.states.size).toBe(1);
    expect(minimized.accepts(["a"])).toBe(true);
    expect(minimized.accepts([])).toBe(true);
  });
});

describe("Minimize with NFA", () => {
  it("should minimize DFA from NFA conversion", () => {
    // NFA for "ends with 'a'"
    const nfa = new NFA(
      new Set(["q0", "q1"]),
      new Set(["a", "b"]),
      new Map([
        [transitionKey("q0", "a"), new Set(["q0", "q1"])],
        [transitionKey("q0", "b"), new Set(["q0"])],
      ]),
      "q0",
      new Set(["q1"]),
    );
    const dfa = nfa.toDfa();
    const minimized = minimize(dfa);

    // The minimal DFA for "ends with 'a'" has exactly 2 states
    expect(minimized.states.size).toBe(2);

    // Verify language
    expect(minimized.accepts(["a"])).toBe(true);
    expect(minimized.accepts(["b", "a"])).toBe(true);
    expect(minimized.accepts(["a", "b", "a"])).toBe(true);
    expect(minimized.accepts(["b"])).toBe(false);
    expect(minimized.accepts(["a", "b"])).toBe(false);
    expect(minimized.accepts([])).toBe(false);
  });

  it("should preserve language exhaustively for strings up to length 3", () => {
    // NFA for "contains 'aa'"
    const nfa = new NFA(
      new Set(["q0", "q1", "q2"]),
      new Set(["a", "b"]),
      new Map([
        [transitionKey("q0", "a"), new Set(["q0", "q1"])],
        [transitionKey("q0", "b"), new Set(["q0"])],
        [transitionKey("q1", "a"), new Set(["q2"])],
        [transitionKey("q2", "a"), new Set(["q2"])],
        [transitionKey("q2", "b"), new Set(["q2"])],
      ]),
      "q0",
      new Set(["q2"]),
    );
    const dfa = nfa.toDfa();
    const minimized = minimize(dfa);

    // Generate all strings up to length 3
    function gen(maxLen: number): string[][] {
      let result: string[][] = [[]];
      for (let i = 0; i < maxLen; i++) {
        const newStrs: string[][] = [];
        for (const s of result) {
          for (const c of ["a", "b"]) {
            newStrs.push([...s, c]);
          }
        }
        result = result.concat(newStrs);
      }
      return result;
    }

    for (const s of gen(3)) {
      expect(nfa.accepts(s)).toBe(minimized.accepts(s));
    }
  });
});

File: code/packages/rust/state-machine/tests/minimize_test.rs
//! Integration tests for DFA minimization (Hopcroft's algorithm).

use std::collections::{HashMap, HashSet};

use state_machine::dfa::DFA;
use state_machine::minimize::minimize;
use state_machine::nfa::NFA;

// ============================================================
// Basic Minimization Tests
// ============================================================

#[test]
fn test_already_minimal() {
    let dfa = DFA::new(
        HashSet::from(["q0".into(), "q1".into()]),
        HashSet::from(["a".into(), "b".into()]),
        HashMap::from([
            (("q0".into(), "a".into()), "q1".into()),
            (("q0".into(), "b".into()), "q0".into()),
            (("q1".into(), "a".into()), "q0".into()),
            (("q1".into(), "b".into()), "q1".into()),
        ]),
        "q0".into(),
        HashSet::from(["q1".into()]),
    )
    .unwrap();
    let minimized = minimize(&dfa);
    assert_eq!(minimized.states().len(), 2);
}

#[test]
fn test_equivalent_states_merged() {
    let dfa = DFA::new(
        HashSet::from(["q0".into(), "q1".into(), "q2".into()]),
        HashSet::from(["a".into(), "b".into()]),
        HashMap::from([
            (("q0".into(), "a".into()), "q1".into()),
            (("q0".into(), "b".into()), "q2".into()),
            (("q1".into(), "a".into()), "q1".into()),
            (("q1".into(), "b".into()), "q1".into()),
            (("q2".into(), "a".into()), "q2".into()),
            (("q2".into(), "b".into()), "q2".into()),
        ]),
        "q0".into(),
        HashSet::from(["q1".into(), "q2".into()]),
    )
    .unwrap();
    let minimized = minimize(&dfa);
    assert_eq!(minimized.states().len(), 2);
}

#[test]
fn test_unreachable_states_removed() {
    let dfa = DFA::new(
        HashSet::from(["q0".into(), "q1".into(), "q_dead".into()]),
        HashSet::from(["a".into()]),
        HashMap::from([
            (("q0".into(), "a".into()), "q1".into()),
            (("q1".into(), "a".into()), "q0".into()),
            (("q_dead".into(), "a".into()), "q_dead".into()),
        ]),
        "q0".into(),
        HashSet::from(["q1".into()]),
    )
    .unwrap();
    let minimized = minimize(&dfa);
    assert_eq!(minimized.states().len(), 2);
}

#[test]
fn test_language_preserved() {
    let dfa = DFA::new(
        HashSet::from(["q0".into(), "q1".into(), "q2".into(), "q3".into()]),
        HashSet::from(["a".into(), "b".into()]),
        HashMap::from([
            (("q0".into(), "a".into()), "q1".into()),
            (("q0".into(), "b".into()), "q2".into()),
            (("q1".into(), "a".into()), "q3".into()),
            (("q1".into(), "b".into()), "q3".into()),
            (("q2".into(), "a".into()), "q3".into()),
            (("q2".into(), "b".into()), "q3".into()),
            (("q3".into(), "a".into()), "q3".into()),
            (("q3".into(), "b".into()), "q3".into()),
        ]),
        "q0".into(),
        HashSet::from(["q1".into(), "q2".into()]),
    )
    .unwrap();
    let minimized = minimize(&dfa);

    let test_inputs: Vec<Vec<&str>> = vec![
        vec!["a"],
        vec!["b"],
        vec!["a", "a"],
        vec!["a", "b"],
        vec!["b", "a"],
        vec![],
    ];
    for events in test_inputs {
        assert_eq!(
            dfa.accepts(&events),
            minimized.accepts(&events),
            "Language mismatch on {:?}",
            events
        );
    }
}

#[test]
fn test_single_state() {
    let dfa = DFA::new(
        HashSet::from(["q0".into()]),
        HashSet::from(["a".into()]),
        HashMap::from([(("q0".into(), "a".into()), "q0".into())]),
        "q0".into(),
        HashSet::from(["q0".into()]),
    )
    .unwrap();
    let minimized = minimize(&dfa);
    assert_eq!(minimized.states().len(), 1);
    assert!(minimized.accepts(&["a"]));
    assert!(minimized.accepts(&[]));
}

// ============================================================
// NFA -> DFA -> Minimize pipeline
// ============================================================

#[test]
fn test_nfa_to_dfa_to_minimized() {
    let nfa = NFA::new(
        HashSet::from(["q0".into(), "q1".into()]),
        HashSet::from(["a".into(), "b".into()]),
        HashMap::from([
            (
                ("q0".into(), "a".into()),
                HashSet::from(["q0".into(), "q1".into()]),
            ),
            (("q0".into(), "b".into()), HashSet::from(["q0".into()])),
        ]),
        "q0".into(),
        HashSet::from(["q1".into()]),
    )
    .unwrap();
    let dfa = nfa.to_dfa();
    let minimized = minimize(&dfa);

    assert_eq!(minimized.states().len(), 2);

    assert!(minimized.accepts(&["a"]));
    assert!(minimized.accepts(&["b", "a"]));
    assert!(minimized.accepts(&["a", "b", "a"]));
    assert!(!minimized.accepts(&["b"]));
    assert!(!minimized.accepts(&["a", "b"]));
    assert!(!minimized.accepts(&[]));
}

#[test]
fn test_minimized_preserves_language_exhaustive() {
    // NFA for "contains 'aa'"
    let nfa = NFA::new(
        HashSet::from(["q0".into(), "q1".into(), "q2".into()]),
        HashSet::from(["a".into(), "b".into()]),
        HashMap::from([
            (
                ("q0".into(), "a".into()),
                HashSet::from(["q0".into(), "q1".into()]),
            ),
            (("q0".into(), "b".into()), HashSet::from(["q0".into()])),
            (("q1".into(), "a".into()), HashSet::from(["q2".into()])),
            (("q2".into(), "a".into()), HashSet::from(["q2".into()])),
            (("q2".into(), "b".into()), HashSet::from(["q2".into()])),
        ]),
        "q0".into(),
        HashSet::from(["q2".into()]),
    )
    .unwrap();
    let dfa = nfa.to_dfa();
    let minimized = minimize(&dfa);

    // Generate all strings up to length 3
    let mut all_strings: Vec<Vec<&str>> = vec![vec![]];
    for _round in 0..3 {
        let mut new_strings: Vec<Vec<&str>> = Vec::new();
        for s in &all_strings {
            for c in &["a", "b"] {
                let mut ns = s.clone();
                ns.push(c);
                new_strings.push(ns);
            }
        }
        all_strings.extend(new_strings);
    }

    for s in &all_strings {
        assert_eq!(nfa.accepts(s), minimized.accepts(s), "Mismatch on {:?}", s);
    }
}

#[test]
fn test_minimize_non_accepting_dfa() {
    // DFA that never accepts
    let dfa = DFA::new(
        HashSet::from(["q0".into(), "q1".into()]),
        HashSet::from(["a".into()]),
        HashMap::from([
            (("q0".into(), "a".into()), "q1".into()),
            (("q1".into(), "a".into()), "q0".into()),
        ]),
        "q0".into(),
        HashSet::new(),
    )
    .unwrap();
    let minimized = minimize(&dfa);
    // Both states behave the same (neither is accepting), should merge to 1
    assert_eq!(minimized.states().len(), 1);
    assert!(!minimized.accepts(&[]));
    assert!(!minimized.accepts(&["a"]));
    assert!(!minimized.accepts(&["a", "a"]));
}
